### PR TITLE
Added piecestore

### PR DESCRIPTION
**Pull request by aleitner**
Store
- [x]  Store Streamed Data
- [x]  Block Data
- [x]  Retrieve Block Data
- [x]  Store Retrieve Data by range
- [x]  Delete Data
- [x] Complete Store test cases 
- [x] Complete Retrieve test cases 
- [x] Complete Delete test cases 

API
Store data
```
func Store(hash string, r io.Reader, length int64, psFileOffset int64, dir string) error
```

Retrieve data
```
func Retrieve(hash string, w io.Writer, length int64, readPosOffset int64, dir string) error
```

Delete data
```
func Delete(hash string, dir string) (error)
```

### Storage
Each shard is saved in a directory structure, the shard hash is split into chunks and the shard is stored in directories with this structure:

```
↳ <2-bytes> directory - first two bytes of shard hash used as the directory name
  ↳ <2-bytes> directory - next two bytes of the shard has used as the directory name
    ↳ <16-bytes> file - the last remaining bytes used for the filename
```

- Maximum total directories is 16 ^ 4 = 65,536
- With 100,000 files in each it would be 6,553,600,000 files total

### Cons
- Can't quickly get statistics on data stored. (Data sizes, Disk Usage)
- Limited by file system

### CODE REVIEW

File: pkg/piecestore/pstore.go [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+	}
+
+	folder1 := string(hash[0:2])
+	folder2 := string(hash[2:4])
+	fileName := string(hash[4:])
+
+	return path.Join(dir, folder1, folder2, fileName), nil
+}
+
+/*
```
Comment (coyle) [inline pkg/piecestore/pstore.go]: This comments formatting seems off

File: pkg/piecestore/pstore.go [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+	if err != nil {
+		return err
+	}
+
+	dataFileSection := fpiece.NewChunk(dataFile, psFileOffset, length)
+
+	// Close when finished
+	defer dataFile.Close()
+
+	buffer := make([]byte, 4096)
```
Comment (coyle) [inline pkg/piecestore/pstore.go]: why 4096 ?

File: pkg/piecestore/pstore.go [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+		} else if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+/*
+	Retrieve
```
Comment (coyle) [inline pkg/piecestore/pstore.go]: This comments formatting seems off

File: pkg/piecestore/pstore.go (hunk 2 above)
Comment (aleitner) [reply pkg/piecestore/pstore.go]: https://stackoverflow.com/a/237495/3830058

File: cmd/piecestore/main.go [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+
+				// Close the file when we are done
+				defer file.Close()
+
+				fileInfo, err := os.Stat(c.Args().Get(1))
+				if err != nil {
+					return err
+				}
+
+				if fileInfo.IsDir() {
```
Comment (jtolds) [inline cmd/piecestore/main.go]: you don't need to do this check. the file is definitely not a directory if os.Open worked

File: pkg/piecestore/pstore.go [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+	returns (error) if 		  failed and nil if successful
+*/
+func Store(hash string, r io.Reader, length int64, psFileOffset int64, dir string) error {
+	if psFileOffset < 0 {
+		return ArgError.New("Offset is less than 0. Must be greater than or equal to 0")
+	}
+	if length < 0 {
+		return ArgError.New("Length is less than 0. Must be greater than or equal to 0")
+	}
+	if dir == "" {
```
Comment (jtolds) [inline pkg/piecestore/pstore.go]: you should put these error checks into PathByHash and have them in one place

File: pkg/piecestore/pstore.go [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+	buffer := make([]byte, 4096)
+	for {
+		// Read data from read stream into buffer
+		n, err := r.Read(buffer)
+		if err == io.EOF {
+			break
+		}
+
+		// Write the buffer to the stream we opened earlier
+		_, err = dataFileChunk.Write(buffer[:n])
```
Comment (jtolds) [inline pkg/piecestore/pstore.go]: you should use io.Copy instead of doing this for loop yourself.

File: cmd/piecestore/main.go (hunk 4 above)
Comment (aleitner) [reply cmd/piecestore/main.go]: https://golang.org/src/os/file.go?s=7402:7439#L23

File: pkg/piecestore/pstore.go (hunk 5 above)
Comment (aleitner) [reply pkg/piecestore/pstore.go]: These checks aren't related to getting the path from the hash though

**Comment by kaloyan-raev** [comment]
@aleitner Travis CI fails due to lint errors:
```
cmd/piecestore/main.go:1::warning: file is not goimported (goimports)
pkg/piecestore/pstore.go:1::warning: file is not goimported (goimports)
pkg/piecestore/pstore.go:12:2:error: cannot find package "github.com/aleitner/FilePiece" in any of: (vet)
cmd/piecestore/main.go:18:5:warning: exported var ArgError should have comment or be unexported (golint)
pkg/piecestore/pstore.go:22:1:warning: comment on exported function PathByHash should be of the form "PathByHash ..." (golint)
pkg/piecestore/pstore.go:35:1:warning: comment on exported function Store should be of the form "Store ..." (golint)
pkg/piecestore/pstore.go:84:1:warning: comment on exported function Retrieve should be of the form "Retrieve ..." (golint)
pkg/piecestore/pstore.go:141:1:warning: comment on exported function Delete should be of the form "Delete ..." (golint)
```
Please, fix them so we can merge this PR.

File: cmd/piecestore/main.go (hunk 4 above)
Comment (jtolds) [reply cmd/piecestore/main.go]: okay so after testing, i still don't think you need to do this check. you are right that the open will succeed but the first read will fail. the practical result is the same.

File: cmd/piecestore/main.go [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+				}
+				if c.Args().Get(1) == "" {
+					return argError.New("Missing file path")
+				}
+				fileInfo, err := os.Stat(c.Args().Get(1))
+				if err != nil {
+					return err
+				}
+
+				if fileInfo.IsDir() != true {
```
Comment (jtolds) [inline cmd/piecestore/main.go]: prefer `!fileInfo.IsDir()`

File: pkg/piecestore/pstore.go [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+
+// Errors
+var (
+	ArgError = errs.Class("argError")
+	FSError  = errs.Class("fsError")
+)
+
+// PathByHash -- creates datapath from hash and dir
+func PathByHash(hash, dir string) (string, error) {
+	if len(hash) < 20 {
```
Comment (jtolds) [inline pkg/piecestore/pstore.go]: you should move most of your hash and dir checking to this method so you don't need to duplicate it in all of the other functions.

overall, without playing "code golf", the less lines of code, the better

File: pkg/piecestore/pstore.go [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+	if readPosOffset >= fileInfo.Size() || readPosOffset < 0 {
+		return 0, ArgError.New("Invalid offset: %v", readPosOffset)
+	}
+
+	// If length less than 0 read the entire file
+	if length <= -1 {
+		length = fileInfo.Size()
+	}
+
+	// If trying to read past the end of the file, just read to the end
```
Comment (jtolds) [inline pkg/piecestore/pstore.go]: doesn't fpiece.NewChunk already handle this?

Comment (jtolds) [reply pkg/piecestore/pstore.go]: also, if the client requested a specific length, i'd prefer to either do what the client said or fail and not try and predict what the client really wanted

File: pkg/piecestore/pstore.go [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+}
+
+// Retrieve -- Retrieve data from pstore directory
+//	hash 					(string)		   Hash of the stored data
+//	w 						(io.Writer)	   Stream that recieves the stored data
+//	length 				(length)		   Amount of data to read. Read all data if -1
+//	readPosOffset	(offset)	   	 Offset of the data that you are reading. Useful for multiple connections to split the data transfer
+//	dir 					(string)		   pstore directory containing all other data stored
+//	returns 			(int64, error) returns err if failed and the number of bytes retrieved if successful
+func Retrieve(hash string, w io.Writer, length int64, readPosOffset int64, dir string) (int64, error) {
```
Comment (jtolds) [inline pkg/piecestore/pstore.go]: it appears the Ranger method went away. we actually need that

**Pull request merged**